build-system/compile/closure-compile.js

```diff
@@ -15,119 +15,119 @@
+  logError(message, err);
+  throw new Error(message);
 }
 
 /**
- * Emits an error to the caller
- *
- * @param {Error} err
+ * Prints an error message when compilation fails
+ * @param {string} message
+ * @param {string} err
  */
-function emitError(err) {
-  err.showStack = false;
-  new EventEmitter().emit('error', err);
+function logError(message, err) {
+  log(`${message}\n` + formatClosureCompilerError(err));
 }
 
 /**
- * Prints an error message when compilation fails
- * @param {string} message
+ * Extracts the actual error message from closure compiler's stdErr by removing
+ * the long preamble that ends with the full command line.
+ * @param {Array<string>} flags
+ * @param {string} stdErr
+ * @return {string}
  */
-function logError(message) {
-  log(`${message}\n` + formatClosureCompilerError(compilerErrors));
+function getErrorText(flags, stdErr) {
+  const lastFlag = flags.slice(-1).pop();
+  return stdErr.split(lastFlag).pop();
 }
 
 /**
- * Normalize the sourcemap file paths before pushing into Closure.
- * Closure don't follow Gulp's normal sourcemap "root" pattern. Gulp considers
- * all files to be relative to the CWD by default, meaning a file `src/foo.js`
- * with a sourcemap alongside points to `src/foo.js`. Closure considers each
- * file relative to the sourcemap. Since the sourcemap for `src/foo.js` "lives"
- * in `src/`, it ends up resolving to `src/src/foo.js`.
- *
- * @param {!NodeJS.WritableStream} closureStream
- * @return {!NodeJS.WritableStream}
+ * Initializes closure compiler with the given set of flags and ensures that the
+ * native compiler is used during compilation.
+ * @param {Array<string>} flags
+ * @return {!Object}
  */
-function makeSourcemapsRelative(closureStream) {
-  const relativeSourceMap = sourcemaps.mapSources((source, file) => {
-    const dir = path.dirname(file.sourceMap.file);
-    return path.relative(dir, source);
-  });
-
-  return pumpify.obj(relativeSourceMap, closureStream);
+function initializeClosure(flags) {
+  const closure = new compiler(flags);
+  closure.JAR_PATH = null;
+  closure.javaPath = getNativeImagePath();
+  return closure;
 }
 
 /**
- * @param {Array<string>} compilerOptions
- * @return {NodeJS.WritableStream}
+ * Runs closure compiler with the given set of flags.
+ * @param {string} outputFilename
+ * @param {!Object} options
+ * @param {Array<string>} flags
+ * @return {Promise<void>}
  */
-function gulpClosureCompile(compilerOptions) {
-  const pluginOptions = {
-    logger: (errors) => (compilerErrors = errors), // Capture compiler errors
-  };
-
-  return makeSourcemapsRelative(
-    closureCompiler.gulp()(compilerOptions, pluginOptions)
-  );
+function runClosure(outputFilename, options, flags) {
+  return new Promise((resolve) => {
+    initializeClosure(flags).run((exitCode, _code, stdErr) => {
+      if (exitCode !== 0) {
+        const err = getErrorText(flags, stdErr);
+        options.typeCheckOnly
```

Comment:
In the case of `handleTypeCheckError` where `options.continueOnError` is true, I don't see `resolve` being called anywhere. 

Which brings me to a nit: It feels dangerous how we call asymmetrically call `resolve` directly within the promise, and yet for the error flows assume the fns will throw an errors to implicitly reject. I think it'd end up being clearer + more maintainable if `handleTypeCheckError` and `handleCompilerError` were actually error getters, and we performed the common handling within runClosure.

Something like this:

```js

const errText = getErrorText(flags, stdErr);
const error  = options.typeCheckOnly ? getTypeCheckError(errText) : getCompilerError(errText, outputFilename);
logError(error, error.message);
if (options.continueOnError) {
  options.errored = true;
  resolve();
  return;
}
reject(error)
```

Reply:
Good point. This was one of the things in the old implementation that was quite goofy. I've updated the handlers to return an `Error` and am now calling `resolve()` or `reject(reason)` under all circumstances. (The change I made is less sweeping than what you suggested, but has the same effect.)